graviton/sim.py

```diff
@@ -0,0 +1,108 @@
+from dataclasses import dataclass
+from typing import Any, Dict, Iterable, List, Optional, Sequence, TypeVar, Union, cast
+
+from algosdk.v2client.algod import AlgodClient
+
+from graviton.abi_strategy import ABICallStrategy
+from graviton.blackbox import DryRunExecutor, DryRunTransactionParams as TxParams
+from graviton.inspector import DryRunProperty as DRProp, DryRunInspector
+from graviton.invariant import Invariant
+from graviton.models import ExecutionMode, PyTypes
+
+# TODO: this will encompass strategies, composed of
+# hypothesis strategies as well as home grown ABIStrategy sub-types
+InputStrategy = Union[Iterable[Sequence[PyTypes]], ABICallStrategy]
+
+
+@dataclass(frozen=True)
+class SimulationResults:
+    succeeded: bool
+    simulate_inspectors: List[DryRunInspector]
+    identities_inspectors: Optional[List[DryRunInspector]] = None
+
+
+class Simulation:
+    """
+    TODO: Slated to become the hypothesis plugin receptor.
+    """
+
+    def __init__(
+        self,
+        algod: AlgodClient,
+        mode: ExecutionMode,
+        simulate_teal: str,
+        predicates: Dict[DRProp, Any],
+        *,
+        abi_method_signature: Optional[str] = None,
+        omit_method_selector: bool = False,
+        validation: bool = True,
+        identities_teal: Optional[str] = None,
+    ):
+        self.simulate_dre: DryRunExecutor = DryRunExecutor(
+            algod,
+            mode,
+            simulate_teal,
+            abi_method_signature=abi_method_signature,
+            omit_method_selector=omit_method_selector,
+            validation=validation,
+        )
+        self.identities_dre: Optional[DryRunExecutor] = None
+        if identities_teal:
+            self.identities_dre = DryRunExecutor(
+                algod,
+                mode,
+                identities_teal,
+                abi_method_signature=abi_method_signature,
+                omit_method_selector=omit_method_selector,
+                validation=validation,
+            )
+        self.predicates: Dict[DRProp, Any] = predicates
+
+        assert self.predicates, "must provide actual predicates to assert with!"
+
+    def run_and_assert(
+        self,
+        inputs: InputStrategy,
+        *,
+        txn_params: Optional[TxParams],
+        verbose: bool = False,
+        msg: str = "",
+    ) -> SimulationResults:
+        assert inputs, "must provide actual inputs to run against!"
```

Comment:
nit: maybe we can lift this assertion by `inputs: InputStrategy`?

Reply:
I'm not sure I understand. In terms of strong typing, this assertion is unnecessary. But since the method is public I still want to ensure that at runtime the user provides actual inputs. But I concede that we could try and make even stronger assertions. For example, on line 85 `cast` is called to make mypy happy, but one could try to make some run-time assertions and validate the shape of `inputs`. I'm counting on the other executor validations would get kicked off on line 89 and hopefully providing enough actionable errors to help a user fix their problems.

Reply:
I see, that makes sense for exported method. Yea type annotation doesn't help too much at runtime tho.